File: src/d_dirimix.c
#include "header.h"


/* static int is_on_simplex(double *x, int dim, int n) */
/* { */
  
/*   int i; */
/*   double sum, *xend; */

/*   for (i = 0; i < n; i++) */
/*     { */
/*       sum = 0; */
      
/*       for (xend = x + dim; x != xend; x++) */
/* 	{ */
/* 	  if (*x < 0) return 0; */
/* 	  sum += *x; */
/* 	} */

/*       if (fabs(sum - 1) > TOL_DIRIREV ) return 0; */
/*     } */
  
/*   return 1; */
/* } */

double ddirimix_point(double *mu, double *nu, double *x,
			  double *w, int dim, int nmu,
			  int take_logs)
 /* Returns the density itself (not the log) */
{
  double density=0;
  int m, i, j ;
  double log_const, log_dens;
  int isOnSimplex;

  isOnSimplex = is_on_simplex(x,  dim) ;
  if( ! isOnSimplex)
    {
      density = take_logs ? ZERO_BMAMEVT : 0 ;
      return density ;
    }
      
  for (m = 0; m < nmu; m++)
    {
      log_const = lgammafn(nu[m]);

      for (i = dim * m; i < dim * (m + 1); i++)
	log_const -= lgammafn(mu[i] * nu[m]);
      /*
	for (ix = 0; ix < nx; ix++)
	{*/
      log_dens = 0;
      for (j = 0; j < dim; j++)
	{
	  log_dens += 
	    (mu[j + dim * m] * nu[m] - 1) * log(x[j]);
	}
	      
      density += w[m] * exp(log_const + log_dens);
    }
      /*}*/
    
 
  return density;
}





 double *compute_ddirimix(double *mu, double *nu, double *x,
			  double *w, int dim, int nmu, int nx,
			  int take_logs)
 /* Returns the densities themselves (not the log) as a vector*/
{
  double *density = calloc(nx+1, sizeof(double));
  if (!density) return NULL;

  int m, i, j, ix;
  double log_const, log_dens;
  int isOnSimplex;
  double  oneOut = 0;
  for (m = 0; m < nmu; m++)
    {
      log_const = lgammafn(nu[m]);

      for (i = dim * m; i < dim * (m + 1); i++)
	log_const -= lgammafn(mu[i] * nu[m]);

      for (ix = 0; ix < nx; ix++)
	{
	  isOnSimplex = is_on_simplex(x + ix * dim,  dim) ;
	  if( ! isOnSimplex)
	    {
	      density[ix]= take_logs ? ZERO_BMAMEVT : 0 ;
	      oneOut = 1;
	    }
	  else
	    {
	      log_dens = 0;
	      
	      for (j = 0; j < dim; j++)
		{
		  log_dens += 
		    (mu[j + dim * m] * nu[m] - 1) * log(x[j + ix * dim]);
		}
	      
	      density[ix] += w[m] * exp(log_const + log_dens);
	    }
	}
    }
  density[nx] = oneOut;
  return density;
}


void d_dirimix(double *x, int *pn, int *pk, int *pp,
	      double *wei, double *mu, double *nu, 
	      int* take_logs, int* return_vector, int *is_error,
	      double *result)
{
  double *density;

  /* if (!is_on_simplex(x, *pp, *pn) || */
  /*       !is_on_simplex(mu, *pp, *pk) ) */
  /*   { */
  /*     *is_error = 1; */
  /*     return; */
  /*   } */
  if(!(density = compute_ddirimix(mu, nu, x, wei, *pp, 
				  *pk, *pn, *take_logs)))
    {
      *is_error=1;
      return;
    }
  int i;

  if (*return_vector)
    {
      for (i = 0; i < *pn; i++)
	  result[i] = *take_logs ? log(density[i]) : density[i];
    }

  else
    {
      *result = 0;
      for (i = 0; i < *pn; i++) 
	*result += log(density[i]);
      
      if (!(*take_logs)) *result = exp(*result);

      if(density[*pn]>0.5)
	*result = *take_logs ? log(ZERO_BMAMEVT) : 0  ;

    }

  free(density);
}

File: R/ddirimix.r
##' Likelihood function (spectral density on the simplex)
##' and angular data  sampler   in the  Dirichlet mixture model.
##'
##'  The spectral probability measure  defined  on the simplex
##' characterizes the
##' dependence structure of multivariate extreme value models.
##' The parameter list for a mixture
##' with  \eqn{k}  components, is made of
##' \describe{
##' \item{Mu}{ The density kernel centers
##' \eqn{\mu_{i,m}, 1\le i \le p, 1\le m \le k}{\mu[1:p,1:k]} :
##' A  \eqn{p*k} matrix,
##' which columns sum to one, and such that \code{Mu \%*\% wei=1},
##' for the moments constraint to be satisfied. 
##' Each column is a Dirichlet kernel center.
##' }
##' \item{wei}{ The weights vector for the kernel densities:
##' A vector of  \eqn{k} positive numbers summing to one.}
##' \item{lnu}{The logarithms of the shape parameters
##' \eqn{nu_m, 1\le m \le k}{\nu[1:k] } for the density kernels:
##' a vector of size \eqn{k}.}
##' }
##' The moments constraint imposes  that the barycenter of the columns in
##' \code{Mu}, with weights \code{wei}, be the center of the simplex.
##' @title Angular density/likelihood function in the Dirichlet Mixture
##' model.
##' @param x An angular data set which may be reduced to a single point: 
##' A \eqn{n*p} matrix or a vector of length \code{p}, where
##' \eqn{p} is the dimension of the sample space and \eqn{n} is
##' the sample size.
##' Each row is a point on the simplex, so that  each row sum to one. 
##' The error tolerance is set to \code{1e-8}
##' in this package.
##' @param par The parameter list for the Dirichlet mixture model.
##'  @param wei Optional. If present, overrides the value of
##' \code{par$wei}.
##' @param Mu Optional. If present, overrides the value of
##' \code{par$Mu}.
##' @param lnu Optional. If present, overrides the value of
##' \code{par$lnu}.
##' @param log Logical: should the density or the likelihood be returned on the log-scale ?
##' @param vectorial Logical: Should a vector of size \eqn{n} or a single value be returned ? 
##' @return \code{ddirimix} returns the likelihood  as a  single number if
##' \code{vectorial ==FALSE}, or as a vector of size
##' \code{nrow(x)} containing the likelihood of each angular data point.
##' If \code{log == TRUE},  the log-likelihood is returned instead.
##' \code{rdirimix} returns a matrix with \code{n} points and
##' \code{p=nrow(Mu)} columns. 
##' @export
ddirimix <-
function(x=c(0.1,0.2,0.7),
           par, #= get("dm.expar.D3k3"),
           wei=par$wei,
           Mu= par$Mu,
           lnu=par$lnu,
           log=FALSE,
           vectorial=FALSE)
  {
    k=length(lnu)
    nu=exp(lnu)
    if(is.vector(x))
      {
        x=matrix(x,nrow=1)
      }
    n=nrow(x)
    p=ncol(x)

    if(vectorial)
      {result=as.double(rep(0,n))}
    else
      {
        result=as.double(0)
      }
    
    C.out=.C("d_dirimix",as.double(t(x)),as.integer(n),as.integer(k),
      as.integer(p),as.double(wei),as.double(Mu),as.double(nu),
      as.integer(log),as.integer(vectorial),is_error=as.integer(0),
      result=result)

    if (C.out$is_error==1)
      {
        stop("in ddirimix: some data or kernels not on the simplex or misbehaved memory allocation.")
      }
    if(C.out$is_error==2)
      stop("in d_dm_dat: misbehaved memory allocation")

    return(C.out$result)
  }




